src/tree/document/c14n.rs

```diff
@@ -0,0 +1,111 @@
+//! Document canonicalization logic
+//!
+use std::ffi::{c_int, c_void, CString};
+use std::os::raw;
+use std::ptr::null_mut;
+
+use crate::tree::c14n::*;
+
+use super::{
+  xmlAllocOutputBuffer, xmlC14NExecute, xmlC14NIsVisibleCallback, xmlChar, xmlNodePtr,
+  xmlOutputBufferClose, xmlOutputBufferPtr, Document,
+};
+
+impl Document {
+  /// Canonicalize a document and return the results.
+  pub fn canonicalize(
+    &self,
+    options: CanonicalizationOptions,
+    callback: Option<(xmlNodePtr, xmlC14NIsVisibleCallback)>,
+  ) -> Result<String, ()> {
+    let document = (*self.0).borrow().doc_ptr;
+
+    let mut ns_list_c = to_xml_string_vec(options.inclusive_ns_prefixes);
+    let inclusive_ns_prefixes = ns_list_c.as_mut_ptr();
+    let with_comments = c_int::from(options.with_comments);
+
+    let (is_visible_callback, user_data) = if let Some((node_ptr, visibility_callback)) = callback {
+      (visibility_callback, node_ptr as *mut _)
+    } else {
+      (None, null_mut())
+    };
+
+    let mode = options.mode.into();
+    unsafe {
+      let c_obuf = create_output_buffer();
+
+      let status = xmlC14NExecute(
+        document,
+        is_visible_callback,
+        user_data,
+        mode,
+        inclusive_ns_prefixes,
+        with_comments,
+        c_obuf,
+      );
+
+      let res = c_obuf_into_output(c_obuf);
+
+      if status < 0 {
+        Err(())
+      } else {
+        Ok(res)
+      }
+    }
+  }
+}
+
+unsafe fn c_obuf_into_output(c_obuf: xmlOutputBufferPtr) -> String {
```

Comment:
My first approach looked like this:

``` rust
let buffer = xmlBufferCreate();
let c14n_res = xmlC14NExecute(
        …,
        xmlOutputBufferCreateBuffer(buffer, ptr::null_mut()),
      );
let result = xmlBufferContent(buffer);
let c_string = CStr::from_ptr(result as *const c_char);
let node_string = c_string.to_string_lossy().
xmlBufferFree(
Ok(node_string)
```

It's shorter and fewer additional functions, but I'm probably missing some edge cases / needed error handling.